firestore/integration_test_internal/src/source_test.cc

```diff
@@ -0,0 +1,523 @@
+  ASSERT_TRUE(snapshot.exists());
+  ASSERT_FALSE(snapshot.metadata().is_from_cache());
+  ASSERT_FALSE(snapshot.metadata().has_pending_writes());
+  ASSERT_EQ(initialData, snapshot.GetData());
+}
+
+TEST_F(SourceTest, TestGetCollectionWhileOnlineWithSourceEqualToServer) {
+  const std::map<std::string, MapFieldValue> initialDocs = {
+      {"doc1", {{"key1", FieldValue::String("value1")}}},
+      {"doc2", {{"key2", FieldValue::String("value2")}}},
+      {"doc3", {{"key3", FieldValue::String("value3")}}}};
+  CollectionReference colRef = Collection(initialDocs);
+
+  Future<QuerySnapshot> future = colRef.Get(Source::kServer);
+  WaitFor(future);
+
+  QuerySnapshot querySnapshot = *future.result();
+  ASSERT_FALSE(querySnapshot.metadata().is_from_cache());
+  ASSERT_FALSE(querySnapshot.metadata().has_pending_writes());
+  ASSERT_EQ(3, querySnapshot.DocumentChanges().size());
+  ASSERT_EQ(initialDocs, QuerySnapshotToMap(querySnapshot));
+}
+
+TEST_F(SourceTest, TestGetDocumentWhileOfflineWithSourceEqualToServer) {
+  MapFieldValue initialData = {{"key", FieldValue::String("value")}};
+  DocumentReference docRef = DocumentWithData(initialData);
+
+  WaitFor(docRef.Get());
+  WaitFor(TestFirestore()->DisableNetwork());
+
+  Future<DocumentSnapshot> future = docRef.Get(Source::kServer);
+  Await(future);
+  EXPECT_EQ(future.status(), FutureStatus::kFutureStatusComplete);
+  EXPECT_EQ(future.error(), Error::kErrorUnavailable);
+}
+
+TEST_F(SourceTest, TestGetCollectionWhileOfflineWithSourceEqualToServer) {
+  const std::map<std::string, MapFieldValue> initialDocs = {
+      {"doc1", {{"key1", FieldValue::String("value1")}}},
+      {"doc2", {{"key2", FieldValue::String("value2")}}},
+      {"doc3", {{"key3", FieldValue::String("value3")}}}};
+  CollectionReference colRef = Collection(initialDocs);
+
+  WaitFor(colRef.Get());
+  WaitFor(TestFirestore()->DisableNetwork());
+
+  Future<QuerySnapshot> future = colRef.Get(Source::kServer);
+  Await(future);
+  EXPECT_EQ(future.status(), FutureStatus::kFutureStatusComplete);
+  EXPECT_EQ(future.error(), Error::kErrorUnavailable);
+}
+
+TEST_F(SourceTest, TestGetDocumentWhileOfflineWithDifferentGetOptions) {
+  MapFieldValue initialData = {{"key", FieldValue::String("value")}};
+  DocumentReference docRef = DocumentWithData(initialData);
+
+  WaitFor(docRef.Get());
+  WaitFor(TestFirestore()->DisableNetwork());
+
+  // Create an initial listener for this query (to attempt to disrupt the
+  // below) and wait for the listener to deliver its initial snapshot before
+  // continuing.
+  std::promise<Error> errorPromise;
+  std::future<Error> errorFuture = errorPromise.get_future();
+
+  docRef.AddSnapshotListener([&errorPromise](const DocumentSnapshot& snapshot,
+                                             Error error_code,
+                                             const std::string& error_message) {
+    errorPromise.set_value(error_code);
+  });
+  errorFuture.wait();
```

Comment:
Nit: please add a check that the value is not an error (I think Android would raise an exception in the error case, so some kind of an error check is warranted). Applies below as well.